tools/crm_mon.c

```diff
@@ -904,6 +907,95 @@ set_fencing_options(int level)
     }
 }
 
+/* Before trying to connect to fencer or cib check for state of
+   pacemakerd - just no sense in trying till pacemakerd has
+   taken care of starting all the sub-processes
+
+   Only noteworthy thing to show here is when pacemakerd is
+   waiting for startup-trigger from SBD.
+ */
+static void
+pacemakerd_event_cb(pcmk_ipc_api_t *pacemakerd_api,
+                    enum pcmk_ipc_event event_type, crm_exit_t status,
+                    void *event_data, void *user_data)
+{
+    pcmk_pacemakerd_api_reply_t *reply = event_data;
+    enum pcmk_pacemakerd_state *state =
+        (enum pcmk_pacemakerd_state *) user_data;
+
+    /* we are just interested in the latest reply */
+    *state = pcmk_pacemakerd_state_invalid;
+
+    switch (event_type) {
+        case pcmk_ipc_event_reply:
+            break;
+
+        default:
+            return;
+    }
+
+    if (status != CRM_EX_OK) {
+        out->err(out, "Bad reply from pacemakerd: %s",
+                 crm_exit_str(status));
+        return;
+    }
+
+    if (reply->reply_type != pcmk_pacemakerd_reply_ping) {
+        out->err(out, "Unknown reply type %d from pacemakerd",
+                 reply->reply_type);
+    } else {
+        if ((reply->data.ping.last_good != (time_t) 0) &&
+            (reply->data.ping.status == pcmk_rc_ok)) {
+            *state = reply->data.ping.state;
+        }
+    }
+}
+
+static int
+pacemakerd_status(void)
+{
+    int rc = pcmk_rc_ok;
+    pcmk_ipc_api_t *pacemakerd_api = NULL;
+    enum pcmk_pacemakerd_state state = pcmk_pacemakerd_state_invalid;
+
+    if (!pcmk_is_set(options.mon_ops, mon_op_cib_native)) {
+        /* we don't need fully functional pacemakerd otherwise */
+        return rc;
+    }
+    rc = pcmk_new_ipc_api(&pacemakerd_api, pcmk_ipc_pacemakerd);
+    if (pacemakerd_api == NULL) {
+        out->err(out, "Could not connect to pacemakerd: %s",
+                 pcmk_rc_str(rc));
+        /* this is unrecoverable so return with rc we have */
+        return rc;
+    }
+    pcmk_register_ipc_callback(pacemakerd_api, pacemakerd_event_cb, (void *) &state);
+    rc = pcmk_connect_ipc(pacemakerd_api, pcmk_ipc_dispatch_poll);
+    if (rc == pcmk_rc_ok) {
+        rc = pcmk_pacemakerd_api_ping(pacemakerd_api, crm_system_name);
+        if (rc == pcmk_rc_ok) {
+            rc = pcmk_poll_ipc(pacemakerd_api, options.reconnect_msec/2);
+            if (rc == pcmk_rc_ok) {
+                pcmk_dispatch_ipc(pacemakerd_api);
+                rc = ENOTCONN;
+                switch (state) {
+                    case pcmk_pacemakerd_state_running:
+                        rc = pcmk_rc_ok;
+                        break;
+                    case pcmk_pacemakerd_state_wait_for_ping:
+                        print_as(output_format ,"Waiting for startup-trigger from SBD ...\n");
+                        break;
```

Comment:
I think it would be worthwhile to print messages for the other states too

Reply:
What I imagine is that the user would see a sequence of messages (they might flash by quickly if everything is going ok, but if it stalls anywhere they'll see that), like

Waiting for cluster to be available on this node ...
Pacemaker daemons starting ...
Waiting for startup-trigger from SBD ...
Waiting for CIB ...

then the actual status

Reply:
And for cluster stops/restarts, it would go back to the beginning

Reply:
With how it is implemented atm it is always going back to the beginning with each retry.
You just don't see the states that aren't printing anything.
Probably showing that Pacemaker-daemons are shutting down is valuable as well although as long as we have a CIB connection we should rather go for that and show what we can get from there.
Good place for the "Waiting for CIB ..." is probably before placing the query - right?
So I guess as long as we have a cib-connection we should skip the ping & printouts.
Let me try out how that looks ...